In [3]:
import numpy as np
import pandas as pd
from sklearn.metrics import roc_auc_score

1. Загрузите данные из файла data-logistic.csv. Это двумерная выбор-
ка, целевая переменная на которой принимает значения -1 или 1.

In [14]:
data = pd.read_csv('data-logistic.csv', header=None)
Y = data.iloc[:, 0]
X1 = data.iloc[:, 1]
X2 = data.iloc[:, 2]

2. Убедитесь, что выше выписаны правильные формулы для гради-
ентного спуска. Обратите внимание, что мы используем полноцен-
ный градиентный спуск, а не его стохастический вариант!

3. Реализуйте градиентный спуск для обычной и L2-регуляризованной
(с коэффициентом регуляризации 10) логистической регрессии. Ис-
пользуйте длину шага k=0.1. В качестве начального приближения
используйте вектор (0, 0).
4. Запустите градиентный спуск и доведите до сходимости (евклидово
расстояние между векторами весов на соседних итерациях долж-
но быть не больше 1e-5). Рекомендуется ограничить сверху число
итераций десятью тысячами.

In [21]:
l = len(Y)
k = 0.1
C = 0
max_it = 10000
eps = 1e-5
w1 = 0.0
w2 = 0.0

for it in range(max_it):
    w1_old, w2_old = w1, w2

    z = w1 * X1 + w2 * X2
    sig = 1 / (1 + np.exp(- Y * z))
    com = Y * (1 - sig)

    w1 += k * (1 / l) * np.sum(X1 * com) - k * C * w1
    w2 += k * (1 / l) * np.sum(X2 * com) - k * C * w2

    if np.sqrt((w1 - w1_old)**2 + (w2 - w2_old)**2) < eps:
        break

a = 1 / (1 + np.exp(-z))

auc1 = roc_auc_score(Y, a)
print("AUC:", round(auc1, 3))

AUC: 0.927


In [22]:
# меняем С = 10
l = len(Y)
k = 0.1
C = 10
max_it = 10000
eps = 1e-5
w1 = 0.0
w2 = 0.0

for it in range(max_it):
    w1_old, w2_old = w1, w2

    z = w1 * X1 + w2 * X2
    sig = 1 / (1 + np.exp(- Y * z))
    com = Y * (1 - sig)

    w1 += k * (1 / l) * np.sum(X1 * com) - k * C * w1
    w2 += k * (1 / l) * np.sum(X2 * com) - k * C * w2

    if np.sqrt((w1 - w1_old)**2 + (w2 - w2_old)**2) < eps:
        break

a = 1 / (1 + np.exp(-z))

auc2 = roc_auc_score(Y, a)
print("AUC:", round(auc2, 3))

AUC: 0.936


5. Какое значение принимает AUC-ROC на обучении без регуляри-
зации и при ее использовании? Эти величины будут ответом на
задание. В качестве ответа приведите два числа через пробел. Об-
ратите внимание, что на вход функции roc_auc_score нужно пода-
вать оценки вероятностей, подсчитанные обученным алгоритмом.
Для этого воспользуйтесь сигмоидной функцией: a(x) = 1/(1 +
exp(−w1x1−w2x2)).

In [23]:
print(round(auc1, 3), round(auc2,3))

0.927 0.936


6. Попробуйте поменять длину шага. Будет ли сходиться алгоритм,
если делать более длинные шаги? Как меняется число итераций
при уменьшении длины шага?

In [57]:
l = len(Y)
k = 0.1
C = 0
max_it = 10000
eps = 1e-5
w1 = 20.0
w2 = 20.0

for it in range(max_it):
    w1_old, w2_old = w1, w2

    z = w1 * X1 + w2 * X2
    sig = 1 / (1 + np.exp(- Y * z))
    com = Y * (1 - sig)

    w1 += k * (1 / l) * np.sum(X1 * com) - k * C * w1
    w2 += k * (1 / l) * np.sum(X2 * com) - k * C * w2

    if np.sqrt((w1 - w1_old)**2 + (w2 - w2_old)**2) < eps:
        break

a = 1 / (1 + np.exp(-z))

auc1 = roc_auc_score(Y, a)
print("AUC:", round(auc1, 3))
print(it)

AUC: 0.927
795


При шаге k=0.1 кол-во итераций равно 243. При увеличении шага k=1 кол-во итераций равно 31 (то есть уменьшается), но алгорим все равно сходится, хоть и с большим шагом. Но уже при k = 2 достигается макс число итераций, что свидетельствует о том, что алгоритм "не сошелся". 

Дальше посмотрели, что будет при уменьшении шага: при к = 0.01 увеличилось количество итераций (1478), при к = 0.001 кол-во итераций 5125, но при этом модель обучилась более качественно. 

И того, можно сделать вывод, что при увеличении шага до адекватного размера будет уменьшаться количество итераций, при этом алгоритм будет сходится и обучение останется качественным, но начиная с определенного k, алгоритм перестает сходится. При уменьшении шага увеличивается количество итераций, и обучение становится более стабильным.

7. Попробуйте менять начальное приближение. Влияет ли оно на что-
нибудь?

В целом при изменении начального приближения просто увеличивается количество итераций.
Например при 10,10 -> кол-во итераций 525, при 20,20 -> 795, и так далее.